k-NN search for `ValueCache`

parikshitbajpai opened the pull request:
Closes #26276

parikshitbajpai committed:
Add k-NN search feature in `ValueCache` (#26276)

parikshitbajpai commented:
@dschwen

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1931170/) on 414c68f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26277/site)

This comment will be updated on new commits.

parikshitbajpai committed:
Fix conditional check for cache size in kNN search (#26276)

parikshitbajpai committed:
Restructure kNN search (#26276)

- Now using `kNearestNeighbors()` function for kNN instead of overloading guess()
- `kNearestNeighbors()` returns a tuple of (key reference, value reference, distance)
- Adds tests for kNN search

dschwen commented on framework/include/utils/ValueCache.h:
```diff
@@ -121,3 +127,29 @@ ValueCache<T>::guess(const std::vector<Real> & in_val, T & out_val, Real & dista
+  // kNN search
+  _kd_tree->findNeighbors(result_set, in_val.data());
+
+  if (result_set.size() < k)
+    mooseWarning(
+        "Requested ", k, " nearest neighbors but current cache size is ", result_set.size());
```
If you emit a warning here you should add a method to query the cache `size()`.

parikshitbajpai replied:
Done. 
I've also moved this warning earlier. Since the `result_set` size is set as `min(_data.size(), k)`, there is no need to search & check based on `result_set.size()`

parikshitbajpai committed:
Add method to get cache size (#26276)

Also removes the warning for cache size smaller than requested number of
neighbors. Since the user cannot act to eliminate the warning, it serves
no purpose.

dschwen commented on framework/include/utils/ValueCache.h:
```diff
@@ -29,7 +30,10 @@ class ValueCache
   ValueCache(std::size_t in_dim, std::size_t max_leaf_size = 10);
 
   void insert(const std::vector<Real> & in_val, const T & out_val);
+  std::size_t cacheSize();
```
let's make that `size()`

dschwen reviewed (approved):
:+1:
